### PR TITLE
Add M5Stack specific builds for esp32 light

### PR DESCRIPTION
**Pull request by andy31415**
This makes the M5Stack display work in case we are building for a m5stack.

A very large chung of this is actually taking existing UIs from all-clusters app and applying to m5stack.
Unfortunately I could see no modularization and created #25551 to fix that. I wanted to have a UI first and then modularize (mainly because I had to go through the entire process of figuring out Kconfig requirements to make all these work ... so now that UI compiles and works, may as well create a PR and not yak shave modularization).

Tested:
   - used `./scripts/build/build_examples.py --enable-flashbundle --target esp32-m5stack-light build` and verified display works
   - used `./scripts/build/build_examples.py --enable-flashbundle --target esp32-devkitc-light build` and flashed on a m5stack (assumed same mcu) and saw it working through logs (tested provision and toggle)

### CODE REVIEW

**Comment by andy31415** [comment]
Created #25551 to fix code duplication for M5Stack UIs ... we shold be able to more easily add UIs for examples than copy and paste (and I also see buttons and LEDs diverging quite a bit).

**Review by tcarmelveilleux — APPROVED** [review]

File: examples/lighting-app/esp32/main/Button.cpp [hunk 1]
```diff
@@ -1,6 +1,7 @@
 /*
  *
  *    Copyright (c) 2022 Project CHIP Authors
```
Comment (cliffamzn) [inline examples/lighting-app/esp32/main/Button.cpp:3]: if we're updating this, should we just make this 2023 too?
```suggestion
 *    Copyright (c) 2022-2023 Project CHIP Authors
```

File: examples/lighting-app/esp32/main/Button.cpp [hunk 2]
```diff
@@ -15,52 +16,102 @@
...
-    gpio_isr_handler_add(static_cast<gpio_num_t>(GPIO_INPUT_IO_0), gpio_isr_handler, (void *) GPIO_INPUT_IO_0);
+    ret = gpio_isr_handler_add(gpioNum, button_isr_handler, (void *) gpioNum);
+    ESP_RETURN_ON_ERROR(ret, TAG, "gpio_isr_handler_add failed: %s", esp_err_to_name(ret));
 
-    ESP_LOGI(TAG, "Button initialized..");
-}
+    mbuttonTimer = xTimerCreate("BtnTmr",               // Just a text name, not used by the RTOS kernel
+                                pdMS_TO_TICKS(50),      // timer period
+                                false,                  // no timer reload (==one-shot)
+                                (void *) (int) gpioNum, // init timer id = gpioNum index
```
Comment (cliffamzn) [inline examples/lighting-app/esp32/main/Button.cpp:105]:
```suggestion
                                (void *) (uint8_t) gpioNum, // init timer id = gpioNum index
```

File: examples/lighting-app/esp32/main/Button.cpp [hunk 3]
```diff
@@ -15,52 +16,102 @@
...
+    return ESP_OK;
+}
+void Button::TimerCallback(TimerHandle_t xTimer)
 {
-    if (button_callback != nullptr)
-    {
-        button_press_handler = button_callback;
-    }
+    // Get the button index of the expired timer and call button event Handler.
+    uint32_t gpio_num = (uint32_t) pvTimerGetTimerID(xTimer);
```
Comment (cliffamzn) [inline examples/lighting-app/esp32/main/Button.cpp:115]: Do GPIOs of the M5 stack actually exceed 255? i don't think they do

```suggestion
    uint8_t gpio_num = (uint8_t) pvTimerGetTimerID(xTimer);
```

File: examples/lighting-app/esp32/main/DeviceCallbacks.cpp [hunk 4]
```diff
@@ -147,3 +148,13 @@ void emberAfOnOffClusterInitCallback(EndpointId endpoint)
     ESP_LOGI(TAG, "emberAfOnOffClusterInitCallback");
     GetAppTask().UpdateClusterState();
 }
+
+void AppDeviceCallbacksDelegate::OnIPv4ConnectivityEstablished()
```
Comment (cliffamzn) [inline examples/lighting-app/esp32/main/DeviceCallbacks.cpp:152]: does the M5 stack only support IPV4?

File: examples/lighting-app/esp32/main/DeviceWithDisplay.cpp [hunk 5]
```diff
@@ -0,0 +1,222 @@
+/*
+ *
+ *    Copyright (c) 2022 Project CHIP Authors
```
Comment (cliffamzn) [inline examples/lighting-app/esp32/main/DeviceWithDisplay.cpp:3]:
```suggestion
 *    Copyright (c) 2023 Project CHIP Authors
```

File: examples/lighting-app/esp32/main/DeviceWithDisplay.cpp [hunk 6]
```diff
@@ -0,0 +1,222 @@
...
+ *    limitations under the License.
+ */
+
+#include "DeviceWithDisplay.h"
+#include "Globals.h"
+#include <app-common/zap-generated/attributes/Accessors.h>
+#include <app-common/zap-generated/ids/Clusters.h>
+#include <setup_payload/QRCodeSetupPayloadGenerator.h>
+
+#if CONFIG_HAVE_DISPLAY
```
Comment (cliffamzn) [inline examples/lighting-app/esp32/main/DeviceWithDisplay.cpp:25]: should this go above line 19?

File: examples/lighting-app/esp32/main/Globals.cpp [hunk 7]
```diff
@@ -0,0 +1,21 @@
+/*
+ *
+ *    Copyright (c) 2022 Project CHIP Authors
```
Comment (cliffamzn) [inline examples/lighting-app/esp32/main/Globals.cpp:3]:
```suggestion
 *    Copyright (c) 2023 Project CHIP Authors
```

File: examples/lighting-app/esp32/main/StatusScreen.cpp [hunk 8]
```diff
@@ -0,0 +1,189 @@
+/*
+ *
+ *    Copyright (c) 2022 Project CHIP Authors
```
Comment (cliffamzn) [inline examples/lighting-app/esp32/main/StatusScreen.cpp:3]:
```suggestion
 *    Copyright (c) 2023 Project CHIP Authors
```

File: examples/lighting-app/esp32/main/include/Button.h [hunk 9]
```diff
@@ -1,6 +1,7 @@
 /*
  *
- *    Copyright (c) 2021 Project CHIP Authors
+ *    Copyright (c) 2022 Project CHIP Authors
```
Comment (cliffamzn) [inline examples/lighting-app/esp32/main/include/Button.h:3]:
```suggestion
 *    Copyright (c) 2021-2023 Project CHIP Authors
```

File: examples/lighting-app/esp32/main/include/Button.h [hunk 10]
```diff
@@ -15,16 +16,74 @@
...
+#define BUTTON_NUMBER 3
 class Button
 {
 public:
-    typedef void (*ButtonPressCallback)(void);
+    /**
+     * @brief
+     *  construct a button
+     */
+    Button();
```
Comment (cliffamzn) [inline examples/lighting-app/esp32/main/include/Button.h:45]: What's the point of providing the empty constructor? Won't it be useless if gpioNum isn't set?

File: examples/lighting-app/esp32/main/Button.cpp (hunk 3 above)
Comment (andy31415) [reply examples/lighting-app/esp32/main/Button.cpp:115]: probably not ... however docs for pvTimerGetTimerId seem to return void* ... I actually only copied this code from existing places so did not review/write it ... it probably is awkward.

File: examples/lighting-app/esp32/main/DeviceCallbacks.cpp (hunk 4 above)
Comment (andy31415) [reply examples/lighting-app/esp32/main/DeviceCallbacks.cpp:152]: When testing it seems it supports both. This was taken from what all-clusters-app does.

**Review by cliffamzn — APPROVED** [review]

File: examples/lighting-app/esp32/sdkconfig_m5stack.defaults [hunk 11]
```diff
@@ -0,0 +1,64 @@
...
+
+#enable BT
+CONFIG_BT_ENABLED=y
+CONFIG_BT_NIMBLE_ENABLED=y
+
+#disable BT connection reattempt
+CONFIG_BT_NIMBLE_ENABLE_CONN_REATTEMPT=n
+
+#enable lwip ipv6 autoconfig
+CONFIG_LWIP_IPV6_AUTOCONFIG=y
```
[the remark this answers is not in the document]
Comment (andy31415) [reply examples/lighting-app/esp32/sdkconfig_m5stack.defaults:39]: I would expect IPv4 and IPv6 configurations to be independent. At the same time, I grabbed this config from existing configurations assuming those were reasonable at some point.

File: examples/lighting-app/esp32/main/Button.cpp (hunk 2 above)
Comment (andy31415) [reply examples/lighting-app/esp32/main/Button.cpp:97]: I believe you are correct, however I am afraid to miss something here - this was copied over from existing Button.cpp so I would prefer to keep it to have the files somewhat in sync for when we want to deduplicate them.

File: examples/lighting-app/esp32/main/Button.cpp (hunk 3 above)
Comment (andy31415) [reply examples/lighting-app/esp32/main/Button.cpp:107]: I believe we should have a cleanup phase once we merge button.cpp (and other) implementations. Not diverging them will make things easier to merge around (even though some things already obviously diverged ... like different LED widget with different capabilities)

File: examples/lighting-app/esp32/main/include/Button.h (hunk 10 above)
Comment (andy31415) [reply examples/lighting-app/esp32/main/include/Button.h:33]: From reading this, it seems there is an Init that can set a gpio number. My guess is empty constructors allows for array definitions.

The interface is awkward though - gpio number should probably be optional (or nullable in some way) and there seems to be a conflict of `button() + init()` or `button(gpio) + init(gpio)` combinations not making sense.

Keeping somewhat as is (but I removed the redundant comments) to have this consistent with other places. Would be nice to have a saner interface though.

**Pull request merged**